cpp/src_prims/metrics/randIndex.h

```diff
@@ -0,0 +1,183 @@
+	uint64_t j = threadIdx.x + blockIdx.x*blockDim.x;
+	uint64_t i = threadIdx.y + blockIdx.y*blockDim.y;
+
+	if(i<size&&j<size){
+		//checking if row <= column to prevent double counting
+		if(j>=i){
+		    return;
+		}
+		//checking if the pair have been classified the same by both the clusters
+		else if(firstClusterArray[i]==firstClusterArray[j]&&secondClusterArray[i]==secondClusterArray[j]){
+		    atomicAdd((unsigned long long int*)a,1);
+		}
+		//checking if the pair have been classified differently by both the clusters
+		else if(firstClusterArray[i]!=firstClusterArray[j]&&secondClusterArray[i]!=secondClusterArray[j]){
+		    atomicAdd((unsigned long long int*)b,1);
+		}
+	}
+}
+
+
+/**
+ * @brief optimized kernel to calculate the values of a and b using block reduce
+ * @param firstClusterArray: the array of classes of type T 
+ * @param secondClusterArray: the array of classes of type T
+ * @param size: the size of the data points
+ * @param a: number of pairs of points that both the clusters have classified the same
+ * @param b: number of pairs of points that both the clusters have classified differently
+ */
+template <typename T, int BLOCK_DIM_X, int BLOCK_DIM_Y>
+__global__ 
+void computeTheNumerator(const T* firstClusterArray,const T* secondClusterArray, uint64_t size, uint64_t *a, uint64_t *b) {
+
+	//calculating the indices of pairs of datapoints compared by the current thread
+	uint64_t j = threadIdx.x + blockIdx.x*blockDim.x;
+	uint64_t i = threadIdx.y + blockIdx.y*blockDim.y;
+
+	//thread-local variables to count a and b
+	uint64_t myA = 0, myB = 0;
+
+	if(i < size && j < size && j<i) {
+
+	//checking if the pair have been classified the same by both the clusters
+	if(firstClusterArray[i]==firstClusterArray[j]&&secondClusterArray[i]==secondClusterArray[j]){
+		++myA;
+	}
+
+	//checking if the pair have been classified differently by both the clusters
+	else if(firstClusterArray[i]!=firstClusterArray[j]&&secondClusterArray[i]!=secondClusterArray[j]){
+		++myB;
+		}			
+	}
+
+	//specialize blockReduce for a 2D block of 1024 threads of type uint64_t	
+	typedef cub::BlockReduce<uint64_t, BLOCK_DIM_X, cub::BLOCK_REDUCE_WARP_REDUCTIONS, BLOCK_DIM_Y> BlockReduce;
+
+	//Allocate shared memory for blockReduce
+	__shared__ typename BlockReduce::TempStorage temp_storage;
+	
+	//summing up thread-local counts specific to a block
+	myA = BlockReduce(temp_storage).Sum(myA);
+	__syncthreads();
+	myB = BlockReduce(temp_storage).Sum(myB);
+	__syncthreads();
+	
+	//executed once per block
+	if(threadIdx.x == 0 && threadIdx.y == 0) {
+		atomicAdd((unsigned long long int*)a, myA);
+		atomicAdd((unsigned long long int*)b, myB);
+	}
+}
+
+
+
+
+/**
+* @brief Function to calculate RandIndex 
+* @param firstClusterArray: the array of classes of type T 
+* @param secondClusterArray: the array of classes of type T
+* @param size: the size of the data points
+* @param randIndex: pointer that stores the randIndex
```

Comment:
This appears to have been changed to a return value, rather than an argument to the function.

Reply:
will do 👍